In [1]:
# import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# import dataframes
path = r'/Users/samira/Desktop/6 Advanced Analytics and Dashboard Design/'
df_times = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'timesData.csv'))
df_shanghai = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'shanghaiData.csv'))
df_cwur = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'cwurData.csv'))

# Rankings

## df_times

In [3]:
df_times

world_rank                        university_name  \
0             1                     Harvard University   
1             2     California Institute of Technology   
2             3  Massachusetts Institute of Technology   
3             4                    Stanford University   
4             5                   Princeton University   
...         ...                                    ...   
2598    601-800                    Yeungnam University   
2599    601-800            Yıldız Technical University   
2600    601-800               Yokohama City University   
2601    601-800           Yokohama National University   
2602    601-800                     Yuan Ze University   

                       country  teaching international  research  citations  \
0     United States of America      99.7          72.4      98.7       98.8   
1     United States of America      97.7          54.6      98.0       99.9   
2     United States of America      97.8          82.3      91.4       99.9   
3     United States of America      98.3          29.5      98.1       99.2   
4     United States of America      90.9          70.3      95.4       99.9   
...                        ...       ...           ...       ...        ...   
2598               South Korea      18.6          24.3      10.9       26.5   
2599                    Turkey      14.5          14.9       7.6       19.3   
2600                     Japan      24.0          16.1      10.2       36.4   
2601                     Japan      20.1          23.3      16.0       13.5   
2602                    Taiwan      16.2          17.7      18.3       28.6   

     income total_score num_students  student_staff_ratio  \
0      34.5        96.1       20,152                  8.9   
1      83.7        96.0        2,243                  6.9   
2      87.5        95.6       11,074                  9.0   
3      64.3        94.3       15,596                  7.8   
4         -        94.2        7,929                  8.4   
...     ...         ...          ...                  ...   
2598   35.4           -       21,958                 15.3   
2599   44.0           -       31,268                 28.7   
2600   37.9           -        4,122                  3.7   
2601   40.4           -       10,117                 12.1   
2602   39.8           -        8,663                 20.6   

     international_students female_male_ratio  year  
0                       25%               NaN  2011  
1                       27%           33 : 67  2011  
2                       33%           37 : 63  2011  
3                       22%           42 : 58  2011  
4                       27%           45 : 55  2011  
...                     ...               ...   ...  
2598                     3%           48 : 52  2016  
2599                     2%           36 : 64  2016  
2600                     3%               NaN  2016  
2601                     8%           28 : 72  2016  
2602                     4%           43 : 57  2016  

[2603 rows x 14 columns]

In [4]:
# check missing values
df_times.isnull().sum()

world_rank                  0
university_name             0
country                     0
teaching                    0
international               0
research                    0
citations                   0
income                      0
total_score                 0
num_students               59
student_staff_ratio        59
international_students     67
female_male_ratio         233
year                        0
dtype: int64

###### Just a few NaN in the 'not so important' variables - which can be ignored for this analysis.

In [5]:
# check duplicates
df_times[df_times.duplicated()]

Empty DataFrame
Columns: [world_rank, university_name, country, teaching, international, research, citations, income, total_score, num_students, student_staff_ratio, international_students, female_male_ratio, year]
Index: []

In [6]:
# check datatypes
df_times.dtypes

world_rank                 object
university_name            object
country                    object
teaching                  float64
international              object
research                  float64
citations                 float64
income                     object
total_score                object
num_students               object
student_staff_ratio       float64
international_students     object
female_male_ratio          object
year                        int64
dtype: object

In [7]:
# check for mixed data
for col in df_times.columns.tolist():
  weird = (df_times[[col]].applymap(type) != df_times[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_times[weird]) > 0:
    print (col)

num_students
international_students
female_male_ratio


In [8]:
# check these data types
df_times['num_students'].dtype

dtype('O')

In [9]:
df_times['international_students'].dtype

dtype('O')

In [10]:
df_times['female_male_ratio'].dtype

dtype('O')

In [11]:
# more cleaning

# replacing the country name
df_times.replace('USA', "United States of America", inplace = True)
df_times.replace('Tanzania', "United Republic of Tanzania", inplace = True)
df_times.replace('Democratic Republic of Congo', "Democratic Republic of the Congo", inplace = True)
df_times.replace('Congo', "Republic of the Congo", inplace = True)
df_times.replace('Lao', "Laos", inplace = True)
df_times.replace('Syrian Arab Republic', "Syria", inplace = True)
df_times.replace('Serbia', "Republic of Serbia", inplace = True)
df_times.replace('Czechia', "Czech Republic", inplace = True)
df_times.replace('UAE', "United Arab Emirates", inplace = True)

# clean df
df_times = df_times[df_times['world_rank'].str.contains("-", case=False) == False]
df_times = df_times[df_times['world_rank'].str.contains("=", case=False) == False]
df_times['world_rank'] = df_times['world_rank'].astype(int)

###### The mixed data types in the variables probably come from the NaNs.

In [12]:
df_times['student_staff_ratio'].dtype

dtype('float64')

In [13]:
# view missing data of 'student_staff_ratio'
student_staff_ratio_nan = df_times[df_times['student_staff_ratio'].isnull() == True]
student_staff_ratio_nan

world_rank                                   university_name  \
45            43                           University of Wisconsin   
81            81                                University of York   
85            86  London School of Economics and Political Science   
111          112                                Bilkent University   
115          115                    University of Hawai’i at Mānoa   
157          158                Georgia Health Sciences University   
246           47  London School of Economics and Political Science   
321          121                                University of York   
361          162              Medical University of South Carolina   
396          197                Georgia Health Sciences University   
640           39  London School of Economics and Political Science   
704          103                                University of York   
790          189              Medical University of South Carolina   
1033          32  London School of Economics and Political Science   
1102         100                                University of York   
1435          34  London School of Economics and Political Science   
1516         113                                University of York   
1825          23  London School of Economics and Political Science   

                       country  teaching international  research  citations  \
45    United States of America      55.5          43.7      64.6       83.4   
81              United Kingdom      47.9          66.6      46.2       81.9   
85              United Kingdom      62.4          99.5      56.2       51.6   
111                     Turkey      34.3          47.7      36.1       95.7   
115   United States of America      38.3          34.2      47.6       81.0   
157   United States of America      67.3          16.3      41.7       48.9   
246             United Kingdom      66.4          96.0      75.4       50.6   
321             United Kingdom      36.7          72.7      38.8       71.6   
361   United States of America      43.9          23.9      38.9       56.6   
396   United States of America      49.0          32.1      14.2       63.5   
640             United Kingdom      70.8          87.2      80.8       66.5   
704             United Kingdom      41.9          71.9      43.9       83.4   
790   United States of America      46.2          27.7      39.6       59.7   
1033            United Kingdom      65.2          81.7      74.2       69.2   
1102            United Kingdom      31.7          73.6      33.2       89.4   
1435            United Kingdom      65.4          87.4      74.2       71.3   
1516            United Kingdom      32.2          75.0      33.8       90.5   
1825            United Kingdom      69.8          93.6      80.6       94.3   

     income total_score num_students  student_staff_ratio  \
45        -        67.0          NaN                  NaN   
81     36.2        59.1          NaN                  NaN   
85     38.4        58.3          NaN                  NaN   
111    32.4        55.4          NaN                  NaN   
115       -        55.2          NaN                  NaN   
157    50.1        50.7          NaN                  NaN   
246    41.1        66.0          NaN                  NaN   
321    33.9        50.4          NaN                  NaN   
361    87.8        45.8          NaN                  NaN   
396    40.7        41.5          NaN                  NaN   
640    42.9        73.1          NaN                  NaN   
704    38.4        57.1          NaN                  NaN   
790    77.1        47.7          NaN                  NaN   
1033   43.6        69.8          NaN                  NaN   
1102   33.3        52.6          NaN                  NaN   
1435   40.6        70.9          NaN                  NaN   
1516   32.2        53.4          NaN                  NaN   
1825   32.2        81.3          NaN                  NaN   

     internat

###### mixed data types?

In [14]:
# further checks
df_times.describe()

world_rank     teaching     research    citations  \
count  1112.000000  1112.000000  1112.000000  1112.000000   
mean     98.201439    51.366097    52.874371    76.696853   
std      58.097290    17.162686    19.925043    15.642485   
min       1.000000    15.900000    13.100000    18.800000   
25%      48.000000    38.575000    37.375000    66.000000   
50%      97.000000    47.900000    48.950000    78.300000   
75%     148.000000    61.725000    65.900000    89.425000   
max     200.000000    99.700000    99.400000   100.000000   

       student_staff_ratio         year  
count          1094.000000  1112.000000  
mean             15.758501  2013.301259  
std               9.064041     1.616209  
min               3.600000  2011.000000  
25%              10.200000  2012.000000  
50%              14.600000  2013.000000  
75%              18.700000  2015.000000  
max              70.400000  2016.000000

## df_shanghai

In [15]:
df_shanghai

world_rank                              university_name national_rank  \
0             1                           Harvard University             1   
1             2                      University of Cambridge             1   
2             3                          Stanford University             2   
3             4           University of California, Berkeley             3   
4             5  Massachusetts Institute of Technology (MIT)             4   
...         ...                                          ...           ...   
4892    401-500                        University of Trieste         11-20   
4893    401-500                       University of Zaragoza          9-13   
4894    401-500                        Utah State University       126-146   
4895    401-500              Vienna University of Technology           4-6   
4896    401-500                       Wake Forest University       126-146   

      total_score  alumni  award   hici     ns    pub   pcp  year  
0           100.0   100.0  100.0  100.0  100.0  100.0  72.4  2005  
1            73.6    99.8   93.4   53.3   56.6   70.9  66.9  2005  
2            73.4    41.1   72.2   88.5   70.9   72.3  65.0  2005  
3            72.8    71.8   76.0   69.4   73.9   72.2  52.7  2005  
4            70.1    74.0   80.6   66.7   65.8   64.3  53.0  2005  
...           ...     ...    ...    ...    ...    ...   ...   ...  
4892          NaN     0.0    0.0    5.0   10.9   25.1  20.1  2015  
4893          NaN     0.0    0.0    7.6    5.1   33.3  13.1  2015  
4894          NaN    13.6    0.0    3.6   10.8   25.1  15.5  2015  
4895          NaN     0.0    0.0    0.0   12.2   28.8  22.9  2015  
4896          NaN     0.0    0.0   14.9    7.5   25.0  11.9  2015  

[4897 rows x 11 columns]

In [16]:
# check missing values
df_shanghai.isnull().sum()

world_rank            0
university_name       1
national_rank         1
total_score        3796
alumni                1
award                 2
hici                  2
ns                   22
pub                   2
pcp                   2
year                  0
dtype: int64

###### 3796 'total_score' missing of 4897 rows (more than 3/4 of the data)

In [17]:
# view missing data
df_shanghai[df_shanghai['university_name'].isnull() == True]

world_rank university_name national_rank  total_score  alumni  award  \
3896         99             NaN           NaN          NaN     NaN    NaN   

      hici  ns  pub  pcp  year  
3896   NaN NaN  NaN  NaN  2013

###### Looks like there is one data entry almost without any data (only world_rank and year).

In [18]:
# delete this almost empty data entry
df_shanghai.dropna(subset = ['university_name'], inplace = True)

In [19]:
# check if deleting worked
df_shanghai[df_shanghai['university_name'].isnull() == True]

Empty DataFrame
Columns: [world_rank, university_name, national_rank, total_score, alumni, award, hici, ns, pub, pcp, year]
Index: []

In [20]:
# view missing data
df_shanghai[df_shanghai['total_score'].isnull() == True]

world_rank                   university_name national_rank  total_score  \
100     101-152                 Aarhus University             2          NaN   
101     101-152  Arizona State University - Tempe         54-71          NaN   
102     101-152        Baylor College of Medicine         54-71          NaN   
103     101-152     Catholic University of Leuven           1-4          NaN   
104     101-152    Catholic University of Louvain           1-4          NaN   
...         ...                               ...           ...          ...   
4892    401-500             University of Trieste         11-20          NaN   
4893    401-500            University of Zaragoza          9-13          NaN   
4894    401-500             Utah State University       126-146          NaN   
4895    401-500   Vienna University of Technology           4-6          NaN   
4896    401-500            Wake Forest University       126-146          NaN   

      alumni  award  hici    ns   pub   pcp  year  
100     15.4   19.3   7.9  22.3  41.6  22.4  2005  
101      0.0   14.4  20.8  26.3  41.9  17.5  2005  
102      0.0    0.0  17.6  34.5  44.0  24.9  2005  
103      0.0    0.0  19.2  16.0  48.7  23.1  2005  
104     14.0   13.9  13.6   8.3  44.7  26.9  2005  
...      ...    ...   ...   ...   ...   ...   ...  
4892     0.0    0.0   5.0  10.9  25.1  20.1  2015  
4893     0.0    0.0   7.6   5.1  33.3  13.1  2015  
4894    13.6    0.0   3.6  10.8  25.1  15.5  2015  
4895     0.0    0.0   0.0  12.2  28.8  22.9  2015  
4896     0.0    0.0  14.9   7.5  25.0  11.9  2015  

[3795 rows x 11 columns]

###### a lot of data is missing in this data set.

In [21]:
# check duplicates
df_shanghai[df_shanghai.duplicated()]

Empty DataFrame
Columns: [world_rank, university_name, national_rank, total_score, alumni, award, hici, ns, pub, pcp, year]
Index: []

In [22]:
# check datatypes
df_shanghai.dtypes

world_rank          object
university_name     object
national_rank       object
total_score        float64
alumni             float64
award              float64
hici               float64
ns                 float64
pub                float64
pcp                float64
year                 int64
dtype: object

In [23]:
# check for mixed data
for col in df_shanghai.columns.tolist():
  weird = (df_shanghai[[col]].applymap(type) != df_shanghai[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_shanghai[weird]) > 0:
    print (col)

In [24]:
# check these data types
df_shanghai['university_name'].dtype

dtype('O')

In [25]:
df_shanghai['national_rank'].dtype

dtype('O')

In [26]:
# further checks
df_shanghai.describe()

total_score       alumni       award         hici           ns  \
count  1101.000000  4896.000000  4895.00000  4895.000000  4875.000000   
mean     36.383470     9.161724     7.69191    16.221491    16.078503   
std      13.557186    14.140636    15.49411    14.382710    12.511529   
min      23.500000     0.000000     0.00000     0.000000     0.000000   
25%      27.400000     0.000000     0.00000     7.300000     8.000000   
50%      31.300000     0.000000     0.00000    12.600000    12.800000   
75%      41.800000    15.600000    13.40000    21.700000    19.800000   
max     100.000000   100.000000   100.00000   100.000000   100.000000   

               pub          pcp         year  
count  4895.000000  4895.000000  4896.000000  
mean     38.254648    21.242329  2009.657884  
std      13.050809     9.254351     3.197546  
min       7.300000     8.300000  2005.000000  
25%      28.900000    15.600000  2007.000000  
50%      36.000000    19.000000  2009.000000  
75%      45.300000    24.500000  2012.000000  
max     100.000000   100.000000  2015.000000

## df_cwur

In [27]:
df_cwur

world_rank                            institution         country  \
0              1                     Harvard University             USA   
1              2  Massachusetts Institute of Technology             USA   
2              3                    Stanford University             USA   
3              4                University of Cambridge  United Kingdom   
4              5     California Institute of Technology             USA   
...          ...                                    ...             ...   
2195         996              University of the Algarve        Portugal   
2196         997                  Alexandria University           Egypt   
2197         998            Federal University of Ceará          Brazil   
2198         999                 University of A Coruña           Spain   
2199        1000        China Pharmaceutical University           China   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
3                 1                    10                 24   
4                 4                     2                 29   
...             ...                   ...                ...   
2195              7                   367                567   
2196              4                   236                566   
2197             18                   367                549   
2198             40                   367                567   
2199             83                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1           NaN   
1                      3            12          4          4           NaN   
2                      5             4          2          2           NaN   
3                      4            16         16         11           NaN   
4                      7            37         22         22           NaN   
...                  ...           ...        ...        ...           ...   
2195                 218           926        845        812         969.0   
2196                 218           997        908        645         981.0   
2197                 218           830        823        812         975.0   
2198                 218           886        974        812         975.0   
2199                 218           861        991        812         981.0   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
3          50   86.17  2012  
4          18   85.21  2012  
...       ...     ...   ...  
2195      816   44.03  2015  
2196      871   44.03  2015  
2197      824   44.03  2015  
2198      651   44.02  2015  
2199      547   44.02  2015  

[2200 rows x 14 columns]

In [28]:
# check missing values
df_cwur.isnull().sum()

world_rank                0
institution               0
country                   0
national_rank             0
quality_of_education      0
alumni_employment         0
quality_of_faculty        0
publications              0
influence                 0
citations                 0
broad_impact            200
patents                   0
score                     0
year                      0
dtype: int64

In [29]:
# view missing data
df_cwur[df_cwur['broad_impact'].isnull() == True]

world_rank                            institution         country  \
0             1                     Harvard University             USA   
1             2  Massachusetts Institute of Technology             USA   
2             3                    Stanford University             USA   
3             4                University of Cambridge  United Kingdom   
4             5     California Institute of Technology             USA   
..          ...                                    ...             ...   
195          96         Australian National University       Australia   
196          97                  University of Alberta          Canada   
197          98                 University of Helsinki         Finland   
198          99     Paris Diderot University - Paris 7          France   
199         100        Georgia Institute of Technology             USA   

     national_rank  quality_of_education  alumni_employment  \
0                1                     7                  9   
1                2                     9                 17   
2                3                    17                 11   
3                1                    10                 24   
4                4                     2                 29   
..             ...                   ...                ...   
195              2                   101                101   
196              4                   101                101   
197              1                    69                101   
198              5                    28                101   
199             57                   101                 85   

     quality_of_faculty  publications  influence  citations  broad_impact  \
0                     1             1          1          1           NaN   
1                     3            12          4          4           NaN   
2                     5             4          2          2           NaN   
3                     4            16         16         11           NaN   
4                     7            37         22         22           NaN   
..                  ...           ...        ...        ...           ...   
195                  43           101        101        101           NaN   
196                 101            68        101         92           NaN   
197                  81            74         79         71           NaN   
198                  72           101         87        101           NaN   
199                 101            97        101         43           NaN   

     patents   score  year  
0          5  100.00  2012  
1          1   91.67  2012  
2         15   89.50  2012  
3         50   86.17  2012  
4         18   85.21  2012  
..       ...     ...   ...  
195      101   44.50  2013  
196       81   44.50  2013  
197      101   44.39  2013  
198      101   44.36  2013  
199       32   44.26  2013  

[200 rows x 14 columns]

In [30]:
# check duplicates
df_cwur[df_cwur.duplicated()]

Empty DataFrame
Columns: [world_rank, institution, country, national_rank, quality_of_education, alumni_employment, quality_of_faculty, publications, influence, citations, broad_impact, patents, score, year]
Index: []

In [31]:
# check datatypes
df_cwur.dtypes

world_rank                int64
institution              object
country                  object
national_rank             int64
quality_of_education      int64
alumni_employment         int64
quality_of_faculty        int64
publications              int64
influence                 int64
citations                 int64
broad_impact            float64
patents                   int64
score                   float64
year                      int64
dtype: object

In [32]:
# check for mixed data
for col in df_cwur.columns.tolist():
  weird = (df_cwur[[col]].applymap(type) != df_cwur[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cwur[weird]) > 0:
    print (col)

In [33]:
# further checks
df_cwur.describe()

world_rank  national_rank  quality_of_education  alumni_employment  \
count  2200.000000    2200.000000           2200.000000        2200.000000   
mean    459.590909      40.278182            275.100455         357.116818   
std     304.320363      51.740870            121.935100         186.779252   
min       1.000000       1.000000              1.000000           1.000000   
25%     175.750000       6.000000            175.750000         175.750000   
50%     450.500000      21.000000            355.000000         450.500000   
75%     725.250000      49.000000            367.000000         478.000000   
max    1000.000000     229.000000            367.000000         567.000000   

       quality_of_faculty  publications    influence    citations  \
count         2200.000000   2200.000000  2200.000000  2200.000000   
mean           178.888182    459.908636   459.797727   413.417273   
std             64.050885    303.760352   303.331822   264.366549   
min              1.000000      1.000000     1.000000     1.000000   
25%            175.750000    175.750000   175.750000   161.000000   
50%            210.000000    450.500000   450.500000   406.000000   
75%            218.000000    725.000000   725.250000   645.000000   
max            218.000000   1000.000000   991.000000   812.000000   

       broad_impact      patents        score         year  
count   2000.000000  2200.000000  2200.000000  2200.000000  
mean     496.699500   433.346364    47.798395  2014.318182  
std      286.919755   273.996525     7.760806     0.762130  
min        1.000000     1.000000    43.360000  2012.000000  
25%      250.500000   170.750000    44.460000  2014.000000  
50%      496.000000   426.000000    45.100000  2014.000000  
75%      741.000000   714.250000    47.545000  2015.000000  
max     1000.000000   871.000000   100.000000  2015.000000

In [34]:
# export
df_times.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'times.csv'))